packages/lib/src/vis/tiles/TiledLayer.tsx

```diff
@@ -1,57 +1,93 @@
-import { Suspense } from 'react';
+import type { ThreeEvent } from '@react-three/fiber';
+import { Suspense, useCallback } from 'react';
 import { LinearFilter, NearestFilter, Vector2 } from 'three';
 import type { Box3 } from 'three';
 
 import type { Size } from '../models';
+import { useAxisSystemContext } from '../shared/AxisSystemProvider';
 import Tile from './Tile';
 import type { TilesApi } from './api';
-import type { ColorMapProps } from './models';
-import { getTileOffsets, scaleToLayer, sortTilesByDistanceTo } from './utils';
+import { useLayerScales } from './hooks';
+import type { ColorMapProps, TileArray } from './models';
+import { useTooltipStore } from './store';
+import { getTileOffsets, scaleBox, sortTilesByDistanceTo } from './utils';
 
 interface Props extends ColorMapProps {
   api: TilesApi;
   layer: number;
   meshSize: Size;
   visibleBox: Box3;
+  topLayer?: boolean;
 }
 
 function TiledLayer(props: Props) {
-  const { api, layer, meshSize, visibleBox, ...colorMapProps } = props;
+  const { api, layer, meshSize, visibleBox, topLayer, ...colorMapProps } =
+    props;
 
   const { baseLayerIndex, numLayers, tileSize } = api;
   const layerSize = api.layerSizes[layer];
+  const { xScale, yScale } = useLayerScales(layerSize, meshSize);
+  const { abscissaConfig, ordinateConfig, worldToData } =
+    useAxisSystemContext();
+  const setTooltipValue = useTooltipStore((state) => state.setTooltipValue);
+
+  const onPointerMove = useCallback(
```

Comment:
If I'm not mistaken, I think you could bring this callback up into `TiledHeatmapMesh` and pass it down as prop only to the top layer.